mls_validation_service/src/handlers.rs

```diff
@@ -81,6 +86,48 @@ impl ValidationApi for ValidationService {
             responses: out,
         }))
     }
+
+    async fn get_association_state(
+        &self,
+        request: Request<GetAssociationStateRequest>,
+    ) -> Result<Response<GetAssociationStateResponse>, Status> {
+        let GetAssociationStateRequest {
+            old_updates,
+            new_updates,
+        } = request.into_inner();
+
+        let conv_proto =
+            |updates: Vec<IdentityUpdateProto>| -> Result<Vec<IdentityUpdate>, Status> {
+                updates
+                    .into_iter()
+                    .map(IdentityUpdate::from_proto)
+                    .collect::<Result<Vec<_>, _>>()
+                    .map_err(|e| Status::new(Code::InvalidArgument, e.to_string()))
+            };
+
+        let to_status = |e: AssociationError| Status::new(Code::InvalidArgument, e.to_string());
+
+        let (mut old_updates, mut new_updates) =
+            (conv_proto(old_updates)?, conv_proto(new_updates)?);
+
+        if old_updates.is_empty() {
+            let new_state = associations::get_state(&new_updates).map_err(to_status)?;
+            return Ok(Response::new(GetAssociationStateResponse {
+                association_state: Some(new_state.clone().into()),
+                state_diff: Some(AssociationStateDiff::from(new_state).into()),
+            }));
+        }
+
+        let old_state = associations::get_state(&old_updates).map_err(to_status)?;
+        old_updates.append(new_updates.as_mut());
+        let new_state = associations::get_state(old_updates).map_err(to_status)?;
```

Comment:
`xmtp_id` has an `apply_update` method specifically designed for this case, so you don't need to re-run the initial updates.

Reply:
nice! I'll add this in